pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,169 @@
+        niFiConfig = NiFiConfig.load(config);
+        if (niFiConfig.getUrl() == null
+                || niFiConfig.getPortName() == null) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+
+        this.start();
+    }
+
+    @Override
+    public void close() throws Exception {
+        this.isRunning = false;
+        if (runnerThread != null) {
+            runnerThread.interrupt();
+            runnerThread.join();
+            runnerThread = null;
+        }
+    }
+
+    protected void start() throws Exception {
+        runnerThread = new Thread(new ReceiveRunnable());
+        runnerThread.setName("Nifi Source Thread");
+        runnerThread.start();
+    }
+
+
+    class ReceiveRunnable implements Runnable {
+
+        public ReceiveRunnable() {
+        }
+
+        @Override
+        public void run() {
+            try {
+                final SiteToSiteClient client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+                try {
+                    while (isRunning) {
+                        final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);
+                        DataPacket dataPacket = transaction.receive();
+                        if (dataPacket == null) {
+                            transaction.confirm();
+                            transaction.complete();
+
+                            // no data available. Wait a bit and try again
+                            try {
+                                Thread.sleep(waitTimeMs);
+                            } catch (InterruptedException e) {
+                            }
+                            continue;
+                        }
+
+                        final List<NiFiDataPacket> dataPackets = new ArrayList<>();
+                        do {
+                            // Read the data into a byte array and wrap it along with the attributes
+                            // into a NiFiDataPacket.
+                            final InputStream inStream = dataPacket.getData();
+                            final byte[] data = new byte[(int) dataPacket.getSize()];
+                            StreamUtils.fillBuffer(inStream, data);
+
+                            final Map<String, String> attributes = dataPacket.getAttributes();
+                            final NiFiDataPacket NiFiDataPacket = new StandardNiFiDataPacket(data, attributes);
+                            dataPackets.add(NiFiDataPacket);
+                            dataPacket = transaction.receive();
+                        } while (dataPacket != null);
```

Comment:
You should add some sort of batching here so that you don't get an OOM exception. We are on the receiving end of a NiFi client that could be sending an unlimited stream of data, so keeping all of these packets in memory isn't a great long term strategy.

Reply:
> You should add some sort of batching here so that you don't get an OOM exception. We are on the receiving end of a NiFi client that could be sending an unlimited stream of data, so keeping all of these packets in memory isn't a great long term strategy.

thanks

Reply:
> You should add some sort of batching here so that you don't get an OOM exception. We are on the receiving end of a NiFi client that could be sending an unlimited stream of data, so keeping all of these packets in memory isn't a great long term strategy.

when we create SiteToSiteClient, use requestBatchCount pulling data from a NiFi instance( the client has the ability to request a particular batch size).

Reply:
> You should add some sort of batching here so that you don't get an OOM exception. We are on the receiving end of a NiFi client that could be sending an unlimited stream of data, so keeping all of these packets in memory isn't a great long term strategy.

when we create SiteToSiteClient, use requestBatchCount pulling data from a NiFi instance( the client has the ability to request a particular batch size).